### DISCUSSION TITLE
Can worker queues be defined via environment variable?

### DISCUSSION THREAD

**Discussion by shawnwall** [celery/celery#9457] [Q&A]
I am using [automatic routing](https://docs.celeryq.dev/en/stable/userguide/routing.html#routing-automatic), along a route definition such as:

`celery_app.conf.task_routes = {"api.tasks.*": {"queue": "somename"}}`

If I run my worker and specify the queues, everything works, ie:

`CMD celery -A api.celery_app worker -l INFO -Q somename`

...but the problem is if I do not specify the queue at command line, the tasks are not enqueued. For my scenario, I'd like the queue name to be defined by an environment variable that would be used at runtime. The only way I've been able to get this working is to manually specify the config of `task_queues` which the celery docs suggest you shouldn't have to do in most scenarios and instead they suggest to use automatic routing.

Is there a way to specify the queue names that the `-Q` arg takes, but do it via an environment variable?

**Comment by gyanu2507** [comment]
Automatic routing only decides which queue the *message* is published to. The worker still only consumes the queues you pass to `-Q`. If you omit `-Q`, it listens on the default queue (`celery`), so those `somename` messages sit unused.

There is no Celery setting that replaces `-Q`. Put the name in the environment and expand it in the process command:

```
celery -A api.celery_app worker -l INFO -Q "$CELERY_QUEUES"
```

Compose / k8s can do the same with `command:` and `${CELERY_QUEUES}`.

`task_queues` is only needed if you want extra queue options (priority, exchange). For a plain named queue, `-Q` plus `task_routes` is enough.
